### PR TITLE
SI-8271 ignore free term values if we can't see them

### PR DESCRIPTION
**Pull request by xeno-by**
Migration of reification from resetAllAttrs to resetLocalAttrs uncovered
a serious bug with reifications of free terms that could emit references
to symbols that don't belong to current lexical scopes. Such symbols can
arise when recursively reifying owners of existentials that can come from
arbitrary places in the program.

This commit ignores values of such free terms (reifying them as ???)
in order to plug an issue that would otherwise be a non-workaroundable
regression in 2.11.0-RC1.

I don't have a very good understanding of this thing, but intuitively
we shouldn't even think of reifying existentials that originate elsewhere.
From what I can see from my brief excursion into typer, a typical approach
of dealing with this is skolemizing existentials that come from elsewhere
and then packing them up by calling existentialAbstraction on created skolems.
Theoretically we could apply the same approach to reified types, but
that will have to remain future work, as that would be too risky for RC1.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym

File: src/compiler/scala/reflect/reify/codegen/GenSymbols.scala [hunk 1]
```diff
@@ -7,6 +7,8 @@ trait GenSymbols {
   self: Reifier =>
 
   import global._
+  private val runDefinitions = currentRun.runDefinitions
+  import runDefinitions._
```
Comment (retronym) [inline src/compiler/scala/reflect/reify/codegen/GenSymbols.scala:11]: Just do conform: `Reifier`-s are transient and can't hang around for more than one run, right? If not, caching `currentRun.runDefintions` would not be correct.

**Comment by retronym** [comment]
Can you explain this further with a test case that is more obviously related to the code being changed? What is the relationship to `WeakTypeTag`? Who could observe the reified `???`?

File: src/compiler/scala/reflect/reify/codegen/GenSymbols.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/reify/codegen/GenSymbols.scala:11]: Yes, they are.

**Comment by xeno-by** [comment]
```
object Macros {
  def foo(c: Context) = bar[List](c)
  def bar[M[X]](d: Context)(implicit tcon: d.WeakTypeTag[M[_]]) = ???
}
```

**Long story**

The problem here happens when we try to materialize `d.WeakTypeTag[M[_]]` for `M = List` by reifying `List[_$1] forSome { type _$1 }`, where `_$1` is owned by `d`.

When reifying non-locatable symbols (roughly speaking, symbols that's not going to be pickled by the pickler phase), we create synthetic symbols exclusively for the reifee (as a part of the materialized typetag/expr code) and then refer to them from the reified type. In order to create such a synthetic symbol, we need an owner, and for that purpose we reify its real world counterpart. Consequently, reification of a non-locatable symbol creates a chain reaction that reifies the entire owner chain until we encounter a locatable owner. 

In our case, this chain reaction started by `_$1` leads to reification of `d`. Reification of `d` leads to creation of something like `newFreeTerm("d", d, <flags>, etc)`. This second argument is the problem, because it can't be referred from the place where reification happens.

**Long story short**

When reifying existentials, we might end up reifying their owners that come from absolutely random places in the program.

**Proposed hotfix**

In order to plug the leak with minimal possible changes, the commit detects whether the value of the free term is going to be visible from the lexical context of reification and inserts `???` if it's not. I.e. `newFreeTerm("d", d, <flags>, etc)` becomes `newFreeTerm("d", ???, <flags>, etc)`.

The impact of this change is going to be minimal. The `???` exception is only going to be tripped if someone decides to inspect the underlying value of that free variable, which doesn't look very likely.

**Comment by xeno-by** [comment]
Actually, values of some free variables are also inspected by toolboxes when they evaluate code, so this might theoretically be a problem. I'll replace `???` with `null`.

**Comment by xeno-by** [comment]
@retronym Just to finish the things with RC1 stuff from myself, we could discuss this pull request as well. It pinpoints and fixes a specific bug, so I think this should be pretty safe, not mentioning that it addresses a regression.

**Comment by xeno-by** [comment]
Ping @retronym

**Comment by retronym** [comment]
@xeno-by Is this change actually needed after the recent reversion to use `resetAllAttrs` in the reifier?

LGTM otherwise as we need to fix the regression. I'd like to revisit this area to find a better fix later on, though.

**Comment by xeno-by** [comment]
Resurgence of resetAllAttrs fixes a regression, as it will simply mask the underlying problem again. Given the potential for regressions if we do merge this PR and the fact that I need to work on nescala slides, let's close now and maybe revisit this issue later during 2.11.x :)

**Comment by adriaanm** [comment]
Ok, next time please also update the fix-by version in Jira to unschedule.
